server/src/main/java/org/elasticsearch/cluster/coordination/StableMasterHealthIndicatorService.java

```diff
@@ -0,0 +1,205 @@
+                impacts.add(
+                    new HealthIndicatorImpact(
+                        3,
+                        "The cluster currently has a master node, but having multiple master node changes in a short time is an indicator "
+                            + "that the cluster is at risk of of not being able to create, delete, or rebalance indices",
+                        List.of(ImpactArea.INGEST)
+                    )
+                );
+                if (includeDetails) {
+                    List<DiscoveryNode> mastersInLast30Minutes = localMasterHistory.getImmutableView();
+                    details.put("current_master", new DiscoveryNodeXContentObject(localMasterHistory.getCurrentMaster()));
+                    details.put("recent_masters", mastersInLast30Minutes.stream().map(DiscoveryNodeXContentObject::new).toList());
+                }
+            } else {
+                logger.trace("The cluster has a stable master node");
+                stableMasterStatus = HealthStatus.GREEN;
+                summary = "The cluster has a stable master node";
+            }
+        } else {
+            stableMasterStatus = HealthStatus.GREEN;
+            summary = "Placeholder summary";
+        }
+
+        return createIndicator(
+            stableMasterStatus,
+            summary,
+            includeDetails ? new SimpleHealthIndicatorDetails(details) : HealthIndicatorDetails.EMPTY,
+            impacts
+        );
+    }
+
+    private long getNumberOfMasterChanges(MasterHistory localMasterHistory) {
+        /*
+         * We want to count all of the transitions to non-null master nodes after the first non-null master node.
+         */
+        return localMasterHistory.getImmutableView().stream().filter(Objects::nonNull).count() - 1;
+    }
+
+    private boolean hasSeenMasterInLast30Seconds() {
+        if (clusterService.state().nodes().getMasterNode() != null) {
+            return true;
+        }
+        return masterHistoryService.getLocalMasterHistory().hasSeenMasterInLastNSeconds(30);
+    }
+
+    /*
+     * If we detect that the same master has gone null 3 or more times, we ask the MasterHistoryService to fetch the master history
+     * as seen from that node so that it is ready in case a health API request comes in.
+     */
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        DiscoveryNode currentMaster = event.state().nodes().getMasterNode();
+        DiscoveryNode previousMaster = event.previousState().nodes().getMasterNode();
+        if (currentMaster == null && previousMaster != null) {
+            if (masterHistoryService.getLocalMasterHistory().hasSameMasterGoneNullNTimes(3)) {
+                DiscoveryNode master = masterHistoryService.getLocalMasterHistory().getMostRecentNonNullMaster();
+                if (master != null) {
+                    masterHistoryService.requestRemoteMasterHistory(master);
```

Comment:
Are we not worried calling this from the cluster state applier thread ? There's some locking which I think happens in the caller thread https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/transport/TcpTransport.java#L351

Reply:
I'll double-check it, but the intention was for this call to be non-blocking. That locking occurs in a different thread once we receive the response asynchronously.

Reply:
Yeah I just confirmed that when making a call to a remote machine this happens on 3 different threads (the first is the clusterApplierService thread, another handles the response to a connection being opened, and the last handles the actual master history reponse. If it's a call to this machine it all happens in one thread (https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/transport/TransportService.java#L440). I'll put in some code to prevent that from happening because it's not useful.

Reply:
This remote call happens at most once per node per cluster change event where the previous master was not null and the current master is null, and the same master has gone null at least 4 times. That ought to be exceedingly rare, and In those cases we actually do want the latest history from the master, right?

Reply:
Also I'd think the thing we'd want to protect from is 50 nodes all hitting the last master (which is already struggling) at once asking for this. We don't have a way to prevent that though do we?

Reply:
Not really, it'd be pretty difficult to coordinate that without involving the master. Don't worry about it, these requests are very lightweight, the master does much heavier O(nodes) stuff anyway.